gpbft/committee.go

```diff
@@ -0,0 +1,56 @@
+package gpbft
+
+import (
+	"fmt"
+	"sync"
+)
+
+var _ CommitteeProvider = (*cachedCommitteeProvider)(nil)
+
+type cachedCommitteeProvider struct {
+	delegate CommitteeProvider
+
+	// mu guards access to committees.
+	mu         sync.Mutex
+	committees map[uint64]*Committee
+}
+
+func newCachedCommitteeProvider(delegate CommitteeProvider) *cachedCommitteeProvider {
+	return &cachedCommitteeProvider{
+		delegate:   delegate,
+		committees: make(map[uint64]*Committee),
+	}
+}
+
+func (c *cachedCommitteeProvider) GetCommittee(instance uint64) (*Committee, error) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	committee, found := c.committees[instance]
+	if !found {
+		var err error
+		switch committee, err = c.delegate.GetCommittee(instance); {
+		case err != nil:
+			return nil, fmt.Errorf("instance %d: %w: %w", instance, ErrValidationNoCommittee, err)
+		case committee == nil:
+			return nil, fmt.Errorf("unexpected nil committee for instance %d", instance)
+		default:
+			if err := committee.PowerTable.Validate(); err != nil {
+				return nil, fmt.Errorf("instance %d: %w: invalid power: %w", instance, ErrValidationNoCommittee, err)
+			}
```

Comment:
This is fine but... it's a bit odd that we're calling this at the caching layer. I'd move this to the host.